### ISSUE TITLE
Error during loadCorrectedReads

### ISSUE THREAD

**Issue by JackyHess** [marbl/canu#1360]
Hi again,

I seem to have run into an issue during the last few steps of error correction. It looks like the correction is finishing fine but there is some conflict in the loadCorrectReads step.

The tail of the log file is as follows:

```-- Found 352 read correction output files.
-- Finished stage 'cor-generateCorrectedReadsCheck', reset canuIteration.
-- Found 352 read correction output files.
-- Finished stage 'cor-generateCorrectedReadsCheck', reset canuIteration.
-- Found 352 read correction output files.
-- Finished stage 'cor-generateCorrectedReadsCheck', reset canuIteration.
--
-- Loading corrected reads into corStore and seqStore.
----------------------------------------
-- Starting command on Tue May 14 09:02:17 2019 with 5938.624 GB free disk space

    cd correction
    /apps/canu/1.8/Linux-amd64/bin/loadCorrectedReads \
      -S ../Tlei.seqStore \
      -C ./Tlei.corStore \
      -L ./2-correction/corjob.files \
    >  ./Tlei.loadCorrectedReads.log \
    2> ./Tlei.loadCorrectedReads.err

-- Finished on Tue May 14 09:02:17 2019 (furiously fast) with 5938.641 GB free disk space
----------------------------------------

ERROR:
ERROR:  Failed with exit code 1.  (rc=256)
ERROR:

ABORT:
ABORT: Canu 1.8
ABORT: Don't panic, but a mostly harmless error occurred and Canu stopped.
ABORT: Try restarting.  If that doesn't work, ask for help.
ABORT:
ABORT:   failed to load corrected reads into store.
ABORT:
```

The content of Tlei.loadCorrectedReads.err shows the following error message:

`sqStoreBlobWriter()-- blob file '../Tlei.seqStore/blobs.0011' already exists.`

I use CANU 1.8 with the following command line:

`canu -correct -p Tlei -d /scratch/hess/TleiHiSen genomeSize=1.2g -pacbio-raw DTG-DNA-499.subreads.bam.fasta maxMemory=320G maxThreads=20 gridEngine=slurm gridOpt
ions='--license=scratch-highio --constraint=array-4core' corMhapFilterThreshold=0.0000000002 corMhapOptions="--threshold 0.80 --num-hashes 512 --num-min-matches 
3 --ordered-sketch-size 1000 --ordered-kmer-size 14 --min-olap-length 2000 --repeat-idf-scale 50" mhapMemory=60g mhapBlockSize=500 ovlMerDistinct=0.975 corMinCov
erage=0 corOutCoverage=200 correctedErrorRate=0.15 gridEngineMemoryOption='--mem=MEMORY' gridEngineThreadsOption='--cpus-per-task=THREADS' gridOptionsJobName=Tle
 stageDirectory=\$TMPDIR`

I've tried rerunning but it seems to get hung up in the same place. Do you know how I might be able to fix this?

Thanks very much for your help!

Jacky

**Comment by brianwalenz** [comment]
What files exist in Tlei.seqStore?
```
% ls -l Tlei.seqStore
```
And dump some of the store metadata for me too:
```
% sqStoreDumpMetaData -S Tlei.seqStore -reads | head
```
I wouldn't mind seeing the canu log for the first run of `loadCorrectedReads`, if you can find it.

**Comment by JackyHess** [comment]
Hi Brian,

Tlei.seqStore looks like this:

```-rw-r--r--. 1 hess login 1073744212 May 10 16:42 blobs.0000
-rw-r--r--. 1 hess login 1073743052 May 10 16:44 blobs.0001
-rw-r--r--. 1 hess login 1073741952 May 10 16:45 blobs.0002
-rw-r--r--. 1 hess login 1073743824 May 10 16:47 blobs.0003
-rw-r--r--. 1 hess login 1073742940 May 10 16:48 blobs.0004
-rw-r--r--. 1 hess login 1073742816 May 10 16:50 blobs.0005
-rw-r--r--. 1 hess login 1073743148 May 10 16:52 blobs.0006
-rw-r--r--. 1 hess login 1073744384 May 10 16:54 blobs.0007
-rw-r--r--. 1 hess login 1073742340 May 10 16:55 blobs.0008
-rw-r--r--. 1 hess login 1073742044 May 10 16:57 blobs.0009
-rw-r--r--. 1 hess login  463233024 May 10 16:58 blobs.0010
-rw-r--r--. 1 hess login 1073744324 May 14 03:14 blobs.0011
-rw-r--r--. 1 hess login 1073743592 May 14 03:15 blobs.0012
-rw-r--r--. 1 hess login 1073745012 May 14 03:16 blobs.0013
-rw-r--r--. 1 hess login 1073745636 May 14 03:17 blobs.0014
-rw-r--r--. 1 hess login 1073744272 May 14 03:18 blobs.0015
-rw-r--r--. 1 hess login 1073751868 May 14 03:19 blobs.0016
-rw-r--r--. 1 hess login 1073745732 May 14 03:20 blobs.0017
-rw-r--r--. 1 hess login 1073749376 May 14 03:21 blobs.0018
-rw-r--r--. 1 hess login 1073742276 May 14 03:22 blobs.0019
-rw-r--r--. 1 hess login 1073746656 May 14 03:23 blobs.0020
-rw-r--r--. 1 hess login 1073747024 May 14 03:24 blobs.0021
-rw-r--r--. 1 hess login 1073743024 May 14 03:25 blobs.0022
-rw-r--r--. 1 hess login 1073743468 May 14 03:26 blobs.0023
-rw-r--r--. 1 hess login  565705532 May 14 03:26 blobs.0024
-rw-r--r--. 1 hess login  117330443 May 10 16:58 errorLog
-rw-r--r--. 1 hess login         88 May 10 16:58 info
-rw-r--r--. 1 hess login        459 May 10 16:58 info.txt
-rw-r--r--. 1 hess login        336 May 10 16:58 libraries
-rw-r--r--. 1 hess login        190 May 10 16:58 libraries.txt
-rw-r--r--. 1 hess login        333 May 10 16:58 load.dat
-rw-r--r--. 1 hess login   28185174 May 10 16:58 readlengths-cor.dat
-rw-r--r--. 1 hess login        373 May 10 16:58 readlengths-cor.gp
-rw-r--r--. 1 hess login       9309 May 10 16:59 readlengths-cor.png
-rw-r--r--. 1 hess login  249951471 May 10 16:58 readNames.txt
-rw-r--r--. 1 hess login  210646600 May 10 16:58 reads
```
and this is the metadata:

``` 
   readID  libraryID     seqLen     rawLen     corLen   clearBgn   clearEnd  segm      byte  part      flags
---------- ---------- ---------- ---------- ---------- ---------- ---------- ----- --------- ----- ----------
         1          1       1208       1208          0          0          0     0         0    0        ---
         2          1       1876       1876          0          0          0     0       384    0        ---
         3          1       2279       2279          0          0          0     0       936    0        ---
         4          1      14515      14515          0          0          0     0      1592    0        ---
         5          1       6681       6681          0          0          0     0      5308    0        ---
         6          1      11339      11339          0          0          0     0      7064    0        ---
         7          1      12652      12652          0          0          0     0      9984    0        ---
         8          1      13270      13270          0          0          0     0     13228    0        ---
```

As for the log files, I can find the output in the canu-logs folder, which looks identical for both runs. Is that what you were after? 

```
Canu v1.8 (+0 commits) r0 .

Current Working Directory:
/scratch/hess/TleiHiSen/correction

Command:
/apps/canu/1.8/Linux-amd64/bin/loadCorrectedReads \
  -S ../Tlei.seqStore \
  -C ./Tlei.corStore \
  -L ./2-correction/corjob.files
```

**Comment by brianwalenz** [comment]
Looks like it failed in the middle of loading the corrected reads.  Delete the `blobs` files from May 14th -- blobs.0011 through blobs.0024 -- and restart.

If you're paranoid (I am), tar up the seqStore directory before restarting.

**Comment by JackyHess** [comment]
Aha! Deleting those files and rerunning revealed the error causing loadCorrectedReads to crash the first time around:

```
-- Loading corrected reads into corStore and seqStore.
----------------------------------------
-- Starting command on Tue May 14 17:52:06 2019 with 5985.513 GB free disk space

    cd correction
    /apps/canu/1.8/Linux-amd64/bin/loadCorrectedReads \
      -S ../Tlei.seqStore \
      -C ./Tlei.corStore \
      -L ./2-correction/corjob.files \
    >  ./Tlei.loadCorrectedReads.log \
    2> ./Tlei.loadCorrectedReads.err
sh: line 5: 277550 Aborted                 /apps/canu/1.8/Linux-amd64/bin/loadCorrectedReads -S ../Tlei.seqStore -C ./Tlei.corStore -L ./2-correction/corjob.files > ./Tlei.loadCorrectedReads.log 2> ./Tlei.loadCorrectedReads.err

-- Finished on Tue May 14 18:13:03 2019 (1257 seconds) with 5956.299 GB free disk space
----------------------------------------

ERROR:
ERROR:  Failed with exit code 134.  (rc=34304)
ERROR:

ABORT:
ABORT: Canu 1.8
ABORT: Don't panic, but a mostly harmless error occurred and Canu stopped.
ABORT: Try restarting.  If that doesn't work, ask for help.
ABORT:
ABORT:   failed to load corrected reads into store.
```
and the tail of Tlei.loadCorrectedReads.err:

```
    12346         0       2-correction/results/0235.cns
    12323         0       2-correction/results/0236.cns
    12282         0       2-correction/results/0237.cns
    12416         0       2-correction/results/0238.cns
    12544         0       2-correction/results/0239.cns
    12434         0       2-correction/results/0240.cns
loadCorrectedReads: stores/loadCorrectedReads.C:163: int main(int, char**): Assertion `read->sqRead_sequen
ceLength(sqRead_corrected) == tig->length()' failed.

Failed with 'Aborted'; backtrace (libbacktrace):
utility/system-stackTrace.C::89 in _Z17AS_UTL_catchCrashiP7siginfoPv()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
stores/loadCorrectedReads.C::163 in main()
(null)::0 in (null)()
(null)::0 in (null)()
```
Thanks again!

**Comment by skoren** [comment]
Ah that might be a bad file. Try removing the failing cns file (0240.cns), re-running correctReads.sh manually (in ```2-correction```, run ```sh correctReads.sh 240```) and then try re-running the loading of the corrected reads.

**Comment by brianwalenz** [comment]
If you still have 0240.cns, could you email it to me?  I'd like to see why it failed.

Warning:  this will have 1/352 of your genome sequence in it.

**Comment by JackyHess** [comment]
I removed 0240.cns and reran the job manually with no indication of failure. After cleaning out the blobs again and rerunning the whole pipeline CANU stumbled over the same block again, with identical error messages.

Brian, i'm emailing you 0240.cns to have a look at. There seems to be something funky about it.

**Comment by brianwalenz** [comment]
Well, I'm stumped.  The format of the file is correct, but 82.7% of the corrected read sequences have a length different than what raw read evidence coordinates say it should be.  Yet the first 239 files seem correct!

1)  I can (tell you how to) disable the check that is failing (it's two lines), which will be enough to get corrected reads output.  With those, a new assembly can be started to generate contigs -- in particular, the new assembly will not know anything about the raw reads or any goofiness in generating them.

2)  On the assumption that it's a bug in generating corrected reads, can you send me the data for a few tens of reads?  This will let me generate corrected reads, hopefully reproducing (and fixing!) the error.
```
% cd TleiHiSen/correction
% falconsense -S ../*seqStore -C ./*corStore -r 1-10 -export a.export
% falconsense -S ../*seqStore -C ./*corStore -r 3584990-3585000 -export b.export
% falconsense -S ../*seqStore -C ./*corStore -r 3585001-3585011 -export c.export
```
This will create three (binary) files, the first with the data to generate the first 10 reads, the second with data for the 10 reads before file 0240.cns, and the third with the first 10 reads in 0240.cns.

3)  On the assumption that there is something bigger broken, can you get permission to send ALL the reads?  I need just the `seqStore` and `corStore` directories.  You can upload to our FTP site.

**Comment by brianwalenz** [comment]
To disable the check, comment out line 163 in file src/stores/loadCorrectedReads.C.  Add two '/'s before the line (or just remove the line entirely):
```
      //assert(read->sqRead_sequenceLength(sqRead_corrected) == tig->length());
```
then recompile (cd src ; make) and restart canu, adding option `-correct` to make it stop after outputting corrected reads.  No need to rerun job 240, but you'll need to clean up the seqStore again.

Then you can start a new canu run using `-pacbio-corrected correctedReads.fasta.gz`.

**Comment by JackyHess** [comment]
Hi Brian,

I wound up implementing option 1 (which worked, thank you!) and purging the assembly files before I could export the data but I have now returned to this and gotten permission to send you the data if you are still interested.

I ran into another problem with this dataset using a different parameterization, I'll post a link when I have made a new issue in case they are related.

Best wishes, 
Jacky

**Comment by brianwalenz** [comment]
Yup, I'd still like to see what's going on here.

**Comment by JackyHess** [comment]
OK, great. They data should be on the FTP server now.

**Comment by skoren** [comment]
I was able to run without error with the latest code and the command:  ```canu -correct -p 'asm' -d asm 'genomeSize=1.2g' -pacbio-raw 'issue1360_reads.fasta.gz' 'corMhapSensitivity=high' 'corMinCoverage=0' 'corOutCoverage=200'```. I'm re-running with 1.8 and the exact command above (without the grid options) but it seems to be fine as well. My only guess is the stage directory caused a collision between jobs, each stage folder should be unique to a single job so should include something like the JOBID in the name, the docs already correctly indicate this (https://canu.readthedocs.io/en/latest/parameter-reference.html?highlight=stageDirectory). Closing.
